src/java.base/share/classes/java/net/URL.java

```diff
@@ -748,23 +803,70 @@ public URL(URL context, String spec, URLStreamHandler handler)
     }
 
     /**
-     * Creates a URL from a URI, as if by invoking {@code uri.toURL()}.
+     * Creates a URL from a URI, as if by invoking {@code uri.toURL()}, but
+     * associating it with the given {@code URLStreamHandler}, if allowed.
+     *
+     * @apiNote
+     * Applications should consider performing additional integrity
+     * checks before constructing a {@code URL} and opening a connection.
+     * See the <a href=#integrity>API note</a> in the class level API
+     * documentation.
+     *
+     * @implSpec The implementation of this method includes calling the {@link
+     * URLStreamHandler#parseURL(URL, String, int, int) parseURL} method on the
+     * selected handler.
+     *
+     * @param uri the {@code URI} from which the returned {@code URL} should
+     *           be built
+     * @param streamHandler a custom protocol stream handler for
+     *                      the returned {@code URL}. Can be {@code null},
+     *                      in which case the default stream handler for
+     *                      the protocol if any, will be used.
+     *
+     * @return a new {@code URL} instance created from the given {@code URI}
+     *   and associated with the given {@code URLStreamHandler}, if any
+     *
+     * @throws NullPointerException if {@code uri} is {@code null}
+     *
+     * @throws IllegalArgumentException if no protocol is specified
+     *         (the {@linkplain URI#getScheme() uri scheme} is {@code null}), or
+     *         if the {@code URLStreamHandler} is not {@code null} and can not be
+     *         set for the given protocol
+     *
+     * @throws  MalformedURLException if an unknown protocol is found,
+     *          or the given URI fails to comply with the specific
+     *          syntax of the associated protocol, or the
+     *          underlying stream handler's {@linkplain
+     *          URLStreamHandler#parseURL(URL, String, int, int)
+     *          parseURL method} throws {@code IllegalArgumentException}
+     *
+     * @throws SecurityException
+     *        if a security manager exists and its
+     *        {@code checkPermission} method doesn't allow
+     *        specifying a stream handler
      *
      * @see java.net.URI#toURL()
+     *
+     * @since 20
      */
-    static URL fromURI(URI uri) throws MalformedURLException {
+    public static URL of(URI uri, URLStreamHandler streamHandler)
```

Comment:
The parameter is named "handler" rather than "streamHandler" in constructors so we should probably keep it the same to avoid any confusion.

Reply:
done